Add CMake ability to check dependencies of libs are correct

nilsdeppe opened the pull request:
## Proposed changes

- Add CMake functions to check that dependencies of libraries are correct
- Add CMake option to check dependencies

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [x] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add option to check spectre library dependencies

nilsdeppe committed:
Add CMake functionality to check target dependencies

nilsdeppe commented:
Rebased on develop

kidder commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+# LIST_OF_ALL_TARGETS: a list of all the targets that might be dependencies
+#                      of the target ${TARGET_NAME}
+# SPECTRE_INCLUDES_OF_FILE:
+#     a list the includes of a file for which to get the dependencies
```
a list of the

kidder commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+# TARGET_LINK_LIBRARIES_COMMAND:
+#     the target_link_libraries command so far
+#
+# Returs by setting `TARGET_LINK_LIBS_COMMAND` in parent scope
```
typo: Returns

kidder reviewed (changes requested):
could you also add something to the dev guide (or wherever you think is appropriate) so developers are aware of these functions

nilsdeppe committed:
Briefly describe def checking functions in dev guide

nilsdeppe commented:
@kidder I pushed a fixup and a new commit adding some documentation to the dev guide. I'm not sure how much more should be said until the checking is fully integrated into CI. I'd be interested in hearing what your thoughts are

kidder commented:
ping @fmahebert

nilsdeppe commented:
Squashed and rebased on develop

fmahebert commented:
Sorry for the delay, I'll get to this early next week 🙂

fmahebert commented on cmake/SpectreCheckDependencies.cmake:
```diff
@@ -0,0 +1,45 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+include(SpectreCheckTargetDependencies)
```
Just a note to check the order of the commits: github is showing me this commit before the commit that adds `SpectreCheckTargetDependencies`. It's presumably github's fault, but should be easy to check.

nilsdeppe replied:
GitHub's fault

fmahebert commented on cmake/SpectreCheckDependencies.cmake:
```diff
@@ -0,0 +1,45 @@
+option(CHECK_ALL_DEPENDENCIES
+  "Check link dependencies of all SpECTRE targets" OFF)
+
+if(CHECK_LIBRARY_DEPENDENCIES OR CHECK_ALL_DEPENDENCIES)
```
For my understanding: why are there two options? They appear to be completely redundant with each other... will they be differentiated in a future PR?

nilsdeppe replied:
My plan is to have one that checks just the libraries and one that checks the libraries and all the libs and executables (python things will need to fit in there somewhere too)

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+    string(REPLACE "\"" "" INCLUDE "${INCLUDE}")
+    list(APPEND SPECTRE_INCLUDES ${INCLUDE})
+  endforeach(INCLUDE ${_RAW_SPECTRE_INCLUDES})
+  # "Return" SPECTRE_INCLUDE by setting it in PARENT_SCOPE
```
typo: SPECTRE_INCLUDE -> SPECTRE_INCLUDES

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+    string(REPLACE ">" "" INCLUDE "${INCLUDE}")
+    list(APPEND SPECTRE_INCLUDES ${INCLUDE})
+  endforeach(INCLUDE ${_RAW_SPECTRE_INCLUDES})
+  # "Return" SPECTRE_INCLUDE by setting it in PARENT_SCOPE
```
typo: SPECTRE_INCLUDE -> SPECTRE_INCLUDES

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+
+# Get the list of 3rd party library includes for the target `TARGET_NAME`.
+#
+# Sets the variable `SPECTRE_INCLUDES` in parent scope to return the result
```
Is it intentional that this function returns TPL includes in the variable SPECTRE_INCLUDES?

nilsdeppe replied:
Yes, but not necessary, so I'll change because I agree it's needlessly confusing :)

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+# LIST_OF_ALL_TARGETS: a list of all the targets that might be dependencies
+#                      of the target ${TARGET_NAME}
+# SPECTRE_INCLUDES_OF_FILE:
+#     a list of the includes of a file for which to get the dependencies
```
I'm struggling to understand what this is saying... so I think a clarification is needed but I can't offer a suggestion

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+# SPECTRE_INCLUDES_OF_FILE:
+#     a list of the includes of a file for which to get the dependencies
+# TARGET_HEADER_FILES:
+#     a list of header files for the target ${TARGET_NAME}
```
Please clarify what headers should be in this list. Is this the complete list of all header files that are included by all files in TARGET_NAME? just the spectre headers? some other subset?

nilsdeppe replied:
I've opted for removing this variable and retrieving it locally from `TARGET_NAME` because I think that makes the code easier to understand

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+# RECURSE_THROUGH_INCLUDES:
+#     if TRUE then any include in any of ${SPECTRE_INCLUDES_OF_FILE}
+#     that is also in the list ${TARGET_HEADER_FILES} has its dependencies
+#     also added.
```
I don't understand this... but I couldn't understand what either of the variables was, so maybe I'll wait for you to clarify those dox before I complain about this one :P

nilsdeppe replied:
I've decided to add a few sentences at the beginning of the documentation to clarify what is happening. Hopefully that'll help (at least with you being able to ask for more clarification) :)

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+
+# Gets the dependencies for a list of includes.
+#
+# Example usage:
```
This example usage is confusing. Many of the arguments don't match their names within the function itself. Are these the names of the variables at the call site? Perhaps describing what the arguments should _be_ (see the comments on the following lines of code) would be more helpful?

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+  set(ABS_HEADER_FILE ${_ABS_HEADER_FILE} PARENT_SCOPE)
+endfunction(_make_header_path_absolute TARGET_NAME HEADER_FILE)
+
+# Gets the dependencies for a list of includes.
```
Which list of includes? can it be arbitrary? should it be the includes from one file? it's unclear why the function is set up this way (in C++ I would have a function to get the deps for 1 include, and another function wrapping it to get the deps from a file, but of course good C++ isn't good cmake). It would be helpful to document how this function is intended to be used.

nilsdeppe replied:
This function gets all dependencies for all includes from one file. You can't really `return` things in CMake and so control flow is horrid, which is why it is written this way.

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+  # expose these includes
+  foreach(SPECTRE_INCLUDE ${SPECTRE_INCLUDES_OF_FILE})
+    # If we are checking header files and the header file is part of
+    # this target, skip checks
```
It's not clear which checks are being skipped, because there's no else branch ... or is this comment misplaced?

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+        continue()
+      endif(${DEP_TARGET} STREQUAL ${TARGET_NAME})
+
+      # Get the header files of DEP_TARG and see if any of the
```
DEP_TARG -> DEP_TARGET ?

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+#     of the target ${TARGET_NAME}
+# LIST_OF_FILES:
+#     a list of the files for which to get the dependencies. Can be
+#     header/tpp or source files.
```
Design question, asking here but could apply to several other functions: why take the target and a list of files? Shouldn't you be checking the deps using all files that are part of the target? Is this a cmake limitation? Or do I misunderstand the design.

nilsdeppe replied:
It means I can recycle code for both SpECTRE and TPL includes. Previously I had it the way you suggested and it resulted in one 250 line function with a lot of copy-paste in for loops.

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+#
+# Since the recursion will cause an infinite loop if there are cyclic includes
+# we limit the recursion depth to 20, at which point an error message is
+# displayed that there may be cyclic dependencies
```
It would be helpful, as you've done in many other cases, to note how this returns by setting something at parent scope.

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,652 @@
+# RECURSE_INCLUDES:
+#     if TRUE then the dependencies of the target's included header files
+#     will also be added. This is used when source files are passed because
+#     the dependencies of the header files must be marked as PUBLIC
```
It would be helpful, as you've done in many other cases, to note how this returns by setting something at parent scope.

fmahebert reviewed (commented):
Some thoughts on documentation, and some confusion, from my first pass through this. Please view these comments as a sign of what confused me, a non-CMake expert... and _not at all_ as any indication that I know what is correct or not correct.

Also, some overall sources of confusion
- I'm often confused, when I see "spectre" in this cmake (e.g. in a function or variable name), if it's a namespacing of sorts or if it's distinguishing a spectre include from a 3rd party lib include.
- I'm also sometimes confused when I try to determine if a function is working with the source C++ headers, or the cmake deps (that are being checked against the C++).

Addressing these sources of confusion is likely outside the scope of this PR, but I thought I'd share a facet of CMake documentation that has room for some clarifications...

nilsdeppe commented:
Rebased on develop and pushed a fixup

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,662 @@
+#     ${TARGET}
+# ALLOWED_EXTRA_TARGETS:
+#     a list of allowed extra dependencies. An example of an extra dependency
+#     is the SpectreFlags target, which supplies no header files and there is
```
"and there" -> "and therefore" ?

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,662 @@
+    _get_deps_for_target(${ARG_TARGET}
+      "${ARG_ALL_TARGETS}" "${SOURCE_FILES}" TRUE)
+    set(_TARGET_PRIVATE_DEPS ${TARGET_DEPS})
+    list(REMOVE_ITEM _TARGET_PRIVATE_DEPS Stl)
```
Can you add a comment to make clear what is happening here? Is it just that the STL is part of the language so therefore doesn't need to be written out in the CMake dependencies?

fmahebert commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,662 @@
+# `INCLUDES_OF_FILE` must be a list of includes from a file (either header
+# or source file) of the target `TARGET_NAME`. If `RECURSE_THROUGH_INCLUDES`
+# is `TRUE` then for each header file in `INCLUDES_OF_FILE` that is also an
+# include of the target `TARGET_NAME`then the header file will also have its
```
This is a very helpful comment, thanks! 👍

But it has a typo: need a space on this line between "`TARGET_NAME`then"

fmahebert reviewed (commented):
Thanks for improving the various comments, I'm finding this easier to understand now! Please squash, including these few remaining things.

nilsdeppe commented:
Rebased on develop and squashed, including the new comments. Thanks for taking the time to make sure this is commented in a way that's actually useful @fmahebert :)

kidder commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,667 @@
+          ${CURRENT_INCLUDE}
+          )
+        # Sets TARGET_DEPS
+        _get_deps_for_includes(${TARGET_NAME} "${LIST_OF_ALL_TARGETS}"
```
I am confused as I thought this function took 4 arguments, not 5...

kidder commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,667 @@
+#     a list of header files for which to get the dependencies.
+# RECURSE_THROUGH_INCLUDES:
+#     if TRUE then any include in any of ${INCLUDES_OF_FILE}
+#     that is also in the list ${TARGET_HEADER_FILES} has its dependencies
```
`TARGET_HEADER_FILES` is an internal variable of the function so it might be nicer to say what it means in terms of the input arguments...

kidder commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -0,0 +1,667 @@
+# LIST_OF_ALLOWED_EXTRA_TARGETS:
+#     a list of allowed extra dependencies. An example of an extra dependency
+#     is the SpectreFlags target, which supplies no header files and there is
+#     not strictly a dependency of ${TARGET_NAME} but can be specified as
```
specified as a

nilsdeppe committed:
Add WrapText.hpp to Utilities headers

nilsdeppe commented:
Rebased on develop, pushed a fixup, and added a commit that adds `WrapText.hpp` to `Utilities/CMakeLists.txt` header files (I found this while running the dep checker over the repo)

kidder reviewed (commented):
you can squash the fixup

nilsdeppe commented:
Okay, squashed. Thanks for the reviews!

kidder commented:
@fmahebert please look at
